### DISCUSSION TITLE
Thread Clarification

### DISCUSSION THREAD

**Discussion by efweber999** [cesanta/mongoose#1901] [Ask questions, get answers] [answered]
Hi,

Regarding the note about threads in the documentation, "NOTE: Since Mongoose's core is not protected against concurrent accesses, make sure that all mg_* API functions are called from the same thread or RTOS task.". I see from forum entries that Mongoose has been made more robust in this regard over time.

If mg_http_listen, and all mg_* API functions except mg_ws_send are run from main, could mg_ws_send be run in a separate thread? No other mg_* functions would be run in that thread, and mg_ws_send would not be run anywhere except that thread.

Thanks

**Comment by scaprile** [comment, marked as answer]
From [the docs](https://mongoose.ws/documentation/#2-minute-integration-guide)
> NOTE: Since Mongoose's core is not protected against concurrent accesses, make sure that all mg_* API functions are called from the same thread or RTOS task.

If you absolutely need multi-threading, take the time to go through the user guide (linked above) and [tutorials](https://mongoose.ws/tutorials/), there is one that shows you how to do [multi-threading](https://mongoose.ws/tutorials/multi-threaded/).

**Reply by efweber999** [reply]
OK, Thank you.
